fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/ComparisonRange.java

```diff
@@ -0,0 +1,166 @@
+ *     </li>
+ * </ul>
+ *
+ * <p>
+ * A {@code ComparisonRange} is an immutable object that provides a variety of methods for producing new range from the
+ * current one and some {@link Comparisons.Comparison} objects. For example, see {@link #tryToAdd(Comparisons.Comparison)}
+ * and {@link #from(Comparisons.Comparison)}.
+ * </p>
+ */
+@API(API.Status.EXPERIMENTAL)
+public class ComparisonRange {
+    public static final ComparisonRange EMPTY = new ComparisonRange();
+
+    @Nullable
+    private final Comparisons.Comparison equalityComparison;
+    @Nullable
+    private final List<Comparisons.Comparison> inequalityComparisons;
+
+    private ComparisonRange() {
+        this.equalityComparison = null;
+        this.inequalityComparisons = null;
+    }
+
+    private ComparisonRange(@Nonnull Comparisons.Comparison equalityComparison) {
+        this.equalityComparison = equalityComparison;
+        this.inequalityComparisons = null;
+    }
+
+    private ComparisonRange(@Nonnull List<Comparisons.Comparison> inequalityComparisons) {
+        this.equalityComparison = null;
+        this.inequalityComparisons = inequalityComparisons;
+    }
+
+    public boolean isEmpty() {
+        return equalityComparison == null && inequalityComparisons == null;
+    }
+
+    public boolean isEquality() {
+        return equalityComparison != null && inequalityComparisons == null;
+    }
+
+    public boolean isInequality() {
+        return equalityComparison == null && inequalityComparisons != null;
+    }
+
+    @Nonnull
+    public Comparisons.Comparison getEqualityComparison() {
+        if (equalityComparison == null) {
+            throw new RecordCoreException("tried to get non-existent equality comparison from ComparisonRange");
+        }
+        return equalityComparison;
+    }
+
+    @Nullable
+    public List<Comparisons.Comparison> getInequalityComparisons() {
+        if (inequalityComparisons == null) {
+            throw new RecordCoreException("tried to get non-existent inequality comparisons from ComparisonRange");
+        }
+        return inequalityComparisons;
+    }
+
+    @Nonnull
+    public Optional<ComparisonRange> tryToAdd(@Nonnull Comparisons.Comparison comparison) {
+        // TODO This isn't actually right becuase of how ranges work...
+        // It appears to be compatible with the AndWithThenPlanner
+        if (isEmpty()) {
+            return from(comparison);
+        } else if (isEquality() && getEqualityComparison().equals(comparison)) {
+            return Optional.of(this);
+        } else if (isInequality() &&
+                ScanComparisons.getComparisonType(comparison).equals(ScanComparisons.ComparisonType.INEQUALITY)) {
+            return Optional.of(new ComparisonRange(ImmutableList.<Comparisons.Comparison>builder()
+                    .addAll(inequalityComparisons)
+                    .add(comparison)
+                    .build()));
```

Comment:
I think I was thinking of this line from the docs:

> For example, the comparison range can include the comparisons `{@code > 8}`, `{@ code < 30}`, and `{@code EQUALS 10}`, but it may optionally simplify this to `{@code > 8}` and `{@code < 30}`.

which seems to imply that you could have both equality and inequality comparisons as long as, like, your equality comparison was contained within the area your inequality comparison cares about. But maybe that will be tidied up with the embettered comparison range merging in general?

Reply:
Yeah, that's what I envision here but haven't implemented yet (and it's non-trivial). For now, the example isn't great, so I've changed `{@code EQUALS 10}` to `{@code < 20}` which expresses the same idea using something that actually works right now.